File: python_files/raylib/incremental_breakout_idle/bricks.py
import pyray as p
from dataclasses import dataclass

from constants import ROWS, COLS, BRICK_WIDTH, BRICK_HEIGHT, BRICK_GAP, MARGIN


@dataclass
class Brick:
    rect: p.Rectangle
    color: p.Color
    thickness: float
    font_size: int
    level: int


class Bricks:
    def __init__(self, level: int) -> None:
        self.list: list[Brick] = []

        for i in range(ROWS):
            for j in range(COLS):
                x: int = j * (BRICK_WIDTH + BRICK_GAP) + MARGIN
                y: int = i * (BRICK_HEIGHT + BRICK_GAP) + MARGIN

                brick: Brick = Brick(
                    rect=p.Rectangle(x, y, BRICK_WIDTH, BRICK_HEIGHT),
                    color=p.BLACK,
                    thickness=5.0,
                    font_size=30,
                    level=level,
                )

                self.list.append(brick)

    def draw(self) -> None:
        for brick in self.list:
            p.draw_rectangle_lines_ex(brick.rect, brick.thickness, brick.color)
            # draw brick level
            text: str = str(brick.level)
            text_width: float = p.measure_text(text, brick.font_size)
            x: int = int(brick.rect.x + brick.rect.width / 2 - text_width/2)
            y: int = int(brick.rect.y + brick.rect.height/2 - brick.font_size/2)

            p.draw_text(text, x, y, brick.font_size, p.BLACK)

File: c_files/raylib/incremental_breakout_idle/bricks.c
#include <raylib.h>
#include <stdio.h>

#include "bricks.h"
#include "constants.h"

#define TEXT_LENGTH 10

void bricks_init(Bricks *bricks, int brick_level)
{
  for (int i = 0; i < ROWS; i++)
  {
    float width = (GetScreenWidth() - MARGIN * 2) / COLS;
    float height = (GetScreenHeight() - MARGIN * 2) / ROWS;
    float gap = 3;

    for (int j = 0; j < COLS; j++)
    {
      int x = j * (width + gap) + MARGIN;
      int y = i * (height + gap) + MARGIN;

      Brick brick;
      brick.rect = (Rectangle){x, y, width, height};
      brick.active = true;
      brick.thickness = 5.0f;
      brick.color = BLACK;
      brick.level = brick_level;

      bricks->list[i * COLS + j] = brick;
    }
  }
}

void bricks_draw(Bricks *bricks)
{
  char text_str[TEXT_LENGTH];

  for (int i = 0; i < ROWS * COLS; i++)
  {
    // draw level of brick
    snprintf(text_str, TEXT_LENGTH, "%d", bricks->list[i].level);
    float font_size = 30;
    float text_width = MeasureText(text_str, font_size);
    Color text_color = BLACK;

    DrawText(text_str, (bricks->list[i].rect.x + bricks->list[i].rect.width / 2 - text_width / 2),
             (bricks->list[i].rect.y + bricks->list[i].rect.height / 2 - font_size / 2),
             font_size,
             text_color);

    DrawRectangleLinesEx(bricks->list[i].rect, bricks->list[i].thickness, bricks->list[i].color);
  }
}